### DISCUSSION TITLE
How to use SqlitePersistentStorage on Android

### DISCUSSION THREAD

**Discussion by FluffyDiscord** [781flyingdutchman/background_downloader#197] [Q&A] [answered]
I am trying to use the SqlitePersistentStorage since I will have thousands of tasks to track and I think the default local_storage will be a bottleneck.

I created FileDownloader in main() with SqlitePersistentStorage, but it's throwing error about "db" not being initialized. When I manually call the initialize() before passing it to FileDownloader, it complains about "db" already being initlized. I have no clue what to do.

**Comment by 781flyingdutchman** [comment]
Hi, sorry this is causing problems. You should call `FileDownloader(persistentStorage: SqlitePersistentStorage());` before the first use of the FileDownloader, and this statement can only be executed once. In the example app therefore it is the first statement in the `initState` of the main screen (it is commented out, but you can see it there), but for a typical app it maybe even better to do it before `runApp`.

If this is causing problems, can you share the code snippet you use, and also the Flutter logs around that time (with the apparent reference to the db not being initialized)?

**Reply by FluffyDiscord** [reply]
My `main.dart` looks like this:

```dart
import "package:background_downloader/background_downloader.dart";

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await FileDownloader(persistentStorage: SqlitePersistentStorage()).trackTasks();

  runApp(const MyApp());
}
```

and this throws:

```
Debug service listening on ws://127.0.0.1:41195/QD38PIx0Dhs=/ws
Syncing files to device sdk gphone64 x86 64...
E/flutter (10026): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: LateInitializationError: Field 'db' has not been initialized.
E/flutter (10026): #0      SqlitePersistentStorage.db (package:background_downloader/src/sqlite_storage.dart)
E/flutter (10026): #1      SqlitePersistentStorage.retrieveAllTaskRecords (package:background_downloader/src/sqlite_storage.dart:186:15)
E/flutter (10026): #2      Database.allRecords (package:background_downloader/src/database.dart:36:39)
E/flutter (10026): #3      BaseDownloader.trackTasks (package:background_downloader/src/base_downloader.dart:321:38)
E/flutter (10026): #4      FileDownloader.trackTasks (package:background_downloader/src/file_downloader.dart:689:23)
E/flutter (10026): #5      main (package:app/main.dart:27:70)
E/flutter (10026): #6      _runMain.<anonymous closure> (dart:ui/hooks.dart:159:23)
E/flutter (10026): #7      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:296:19)
E/flutter (10026): #8      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:189:12)
E/flutter (10026): 
```

**Comment by 781flyingdutchman** [comment, marked as answer]
I suggest separating the trackTasks call, and running it with a slight delay:
`Future.delayed(const Duration(milliseconds: 50), (_) => FileDownloader().trackTasks())`

The database initialization is async, but not awaited so I think what's happening here is that you call trackTasks when the DB initialization is still running. I'll look into a more elegant solution, but please try this for now

**Reply by FluffyDiscord** [reply]
I didn't think of that :D Thanks

**Comment by 781flyingdutchman** [comment]
V8.0.0 adds a `ready` getter for `FileDownloader` and `trackTasks` waits for the downloader to be ready before manipulating the database, so I believe this should now be fixed, and your original (more elegant!) code `await FileDownloader(persistentStorage: SqlitePersistentStorage()).trackTasks();` should work.

However, `SqlitePersistentStorage` has moved out of the background_downloader into a separate package `background_downloader_sql` which is a small extension package with the `sqflite` dependency, so you need to add this to your pubspec.yaml alongside background_downloader. The reason is that `sqflite` adds quite a bit of size to an app, and the default downloader does not require it, so it adds overhead for many, to benefit only those who use  `SqlitePersistentStorage`.
